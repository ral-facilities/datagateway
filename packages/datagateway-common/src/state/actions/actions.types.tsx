import {
  DownloadCart,
  Entity,
  Filter,
  FiltersType,
  Order,
  QueryParams,
  SortType,
  ViewsType,
  MicroFrontendId,
  Datafile,
  Dataset,
  Instrument,
  Investigation,
} from '../../app.types';
import { DlsDatasetDetailsPanelTab } from '../../detailsPanels/dls/datasetDetailsPanel.component';
import { DlsVisitDetailsPanelTab } from '../../detailsPanels/dls/visitDetailsPanel.component';
import type { IsisDatafileDetailsPanelTab } from '../../detailsPanels/isis/datafileDetailsPanel.component';
import type { IsisDatasetDetailsPanelTab } from '../../detailsPanels/isis/datasetDetailsPanel.component';
import type { IsisInstrumentDetailsPanelTab } from '../../detailsPanels/isis/instrumentDetailsPanel.component';
import type { IsisInvestigationDetailsPanelTab } from '../../detailsPanels/isis/investigationDetailsPanel.component';

// parent app actions
export const CustomFrontendMessageType = `${MicroFrontendId}:api`;
export const NotificationType = `${CustomFrontendMessageType}:notification`;
export const InvalidateTokenType = `${CustomFrontendMessageType}:invalidate_token`;
export const RegisterRouteType = `${CustomFrontendMessageType}:register_route`;
export const RequestPluginRerenderType = `${CustomFrontendMessageType}:plugin_rerender`;
export const SendThemeOptionsType = `${CustomFrontendMessageType}:send_themeoptions`;
export const BroadcastSignOutType = `${CustomFrontendMessageType}:signout`;

// internal actions
export const ConfigureFacilityNameType =
  'datagateway_common:configure_facility_name';
export const ConfigureURLsType = 'datagateway_common:configure_urls';

export const SortTableType = 'datagateway_common:sort_table';
export const FilterTableType = 'datagateway_common:filter_table';
export const SortUpdateType = 'datagateway_common:sort_update';
export const FilterUpdateType = 'datagateway_common:filter_update';
export const PageUpdateType = 'datagateway_common:page_update';
export const ResultsUpdateType = 'datagateway_common:results_update';

export const UpdateFiltersType = 'datagateway_common:update_filters';
export const UpdateSortType = 'datagateway_common:update_sort';
export const UpdateQueryType = 'datagateway_common:update_query';

export const UpdateViewType = 'datagateway_common:update_view';
export const UpdateSearchType = 'datagateway_common:update_search';
export const UpdatePageType = 'datagateway_common:update_page';
export const UpdateResultsType = 'datagateway_common:update_results';

export const UpdateSaveViewType = 'datagateway_common:update_save_view';

export const ClearTableType = 'datagateway_common:clear_table';
export const ClearDataType = 'datagateway_common:clear_data';

export const FetchInvestigationsRequestType =
  'datagateway_common:fetch_investigations_request';
export const FetchInvestigationsFailureType =
  'datagateway_common:fetch_investigations_failure';
export const FetchInvestigationsSuccessType =
  'datagateway_common:fetch_investigations_success';

export const FetchInvestigationDetailsRequestType =
  'datagateway_common:fetch_investigation_details_request';
export const FetchInvestigationDetailsFailureType =
  'datagateway_common:fetch_investigation_details_failure';
export const FetchInvestigationDetailsSuccessType =
  'datagateway_common:fetch_investigation_details_success';

export const FetchInvestigationCountRequestType =
  'datagateway_common:fetch_investigation_count_request';
export const FetchInvestigationCountFailureType =
  'datagateway_common:fetch_investigation_count_failure';
export const FetchInvestigationCountSuccessType =
  'datagateway_common:fetch_investigation_count_success';

export const FetchInvestigationSizeRequestType =
  'datagateway_common:fetch_investigation_size_request';
export const FetchInvestigationSizeFailureType =
  'datagateway_common:fetch_investigation_size_failure';
export const FetchInvestigationSizeSuccessType =
  'datagateway_common:fetch_investigation_size_success';

export const FetchInvestigationDatasetsCountRequestType =
  'datagateway_common:fetch_investigation_datasets_count_request';
export const FetchInvestigationDatasetsCountFailureType =
  'datagateway_common:fetch_investigation_datasets_count_failure';
export const FetchInvestigationDatasetsCountSuccessType =
  'datagateway_common:fetch_investigation_datasets_count_success';

export const FetchDatasetsRequestType =
  'datagateway_common:fetch_datasets_request';
export const FetchDatasetsFailureType =
  'datagateway_common:fetch_datasets_failure';
export const FetchDatasetsSuccessType =
  'datagateway_common:fetch_datasets_success';

export const FetchDatasetDetailsRequestType =
  'datagateway_common:fetch_dataset_details_request';
export const FetchDatasetDetailsFailureType =
  'datagateway_common:fetch_dataset_details_failure';
export const FetchDatasetDetailsSuccessType =
  'datagateway_common:fetch_dataset_details_success';

export const FetchDatasetCountRequestType =
  'datagateway_common:fetch_dataset_count_request';
export const FetchDatasetCountFailureType =
  'datagateway_common:fetch_dataset_count_failure';
export const FetchDatasetCountSuccessType =
  'datagateway_common:fetch_dataset_count_success';

export const FetchDatasetSizeRequestType =
  'datagateway_common:fetch_dataset_size_request';
export const FetchDatasetSizeFailureType =
  'datagateway_common:fetch_dataset_size_failure';
export const FetchDatasetSizeSuccessType =
  'datagateway_common:fetch_dataset_size_success';

export const FetchDatasetDatafilesCountRequestType =
  'datagateway_common:fetch_dataset_datafiles_count_request';
export const FetchDatasetDatafilesCountFailureType =
  'datagateway_common:fetch_dataset_datafiles_count_failure';
export const FetchDatasetDatafilesCountSuccessType =
  'datagateway_common:fetch_dataset_datafiles_count_success';

export const DownloadDatasetRequestType =
  'datagateway_common:download_dataset_request';
export const DownloadDatasetFailureType =
  'datagateway_common:download_dataset_failure';
export const DownloadDatasetSuccessType =
  'datagateway_common:download_dataset_success';

export const FetchDatafilesRequestType =
  'datagateway_common:fetch_datafiles_request';
export const FetchDatafilesFailureType =
  'datagateway_common:fetch_datafiles_failure';
export const FetchDatafilesSuccessType =
  'datagateway_common:fetch_datafiles_success';

export const FetchDatafileDetailsRequestType =
  'datagateway_common:fetch_datafile_details_request';
export const FetchDatafileDetailsFailureType =
  'datagateway_common:fetch_datafile_details_failure';
export const FetchDatafileDetailsSuccessType =
  'datagateway_common:fetch_datafile_details_success';

export const FetchDatafileCountRequestType =
  'datagateway_common:fetch_datafile_count_request';
export const FetchDatafileCountFailureType =
  'datagateway_common:fetch_datafile_count_failure';
export const FetchDatafileCountSuccessType =
  'datagateway_common:fetch_datafile_count_success';

export const DownloadDatafileRequestType =
  'datagateway_common:download_datafile_request';
export const DownloadDatafileFailureType =
  'datagateway_common:download_datafile_failure';
export const DownloadDatafileSuccessType =
  'datagateway_common:download_datafile_success';

export const FetchInstrumentsRequestType =
  'datagateway_common:fetch_instruments_request';
export const FetchInstrumentsFailureType =
  'datagateway_common:fetch_instruments_failure';
export const FetchInstrumentsSuccessType =
  'datagateway_common:fetch_instruments_success';

export const FetchInstrumentDetailsRequestType =
  'datagateway_common:fetch_instrument_details_request';
export const FetchInstrumentDetailsFailureType =
  'datagateway_common:fetch_instrument_details_failure';
export const FetchInstrumentDetailsSuccessType =
  'datagateway_common:fetch_instrument_details_success';

export const FetchInstrumentCountRequestType =
  'datagateway_common:fetch_instrument_count_request';
export const FetchInstrumentCountFailureType =
  'datagateway_common:fetch_instrument_count_failure';
export const FetchInstrumentCountSuccessType =
  'datagateway_common:fetch_instrument_count_success';

export const FetchFacilityCyclesRequestType =
  'datagateway_common:fetch_facility_cycles_request';
export const FetchFacilityCyclesFailureType =
  'datagateway_common:fetch_facility_cycles_failure';
export const FetchFacilityCyclesSuccessType =
  'datagateway_common:fetch_facility_cycles_success';

export const FetchFacilityCycleCountRequestType =
  'datagateway_common:fetch_facility_cycle_count_request';
export const FetchFacilityCycleCountFailureType =
  'datagateway_common:fetch_facility_cycle_count_failure';
export const FetchFacilityCycleCountSuccessType =
  'datagateway_common:fetch_facility_cycle_count_success';

export const FetchStudiesRequestType =
  'datagateway_common:fetch_studies_request';
export const FetchStudiesFailureType =
  'datagateway_common:fetch_studies_failure';
export const FetchStudiesSuccessType =
  'datagateway_common:fetch_studies_success';

export const FetchStudyCountRequestType =
  'datagateway_common:fetch_study_count_request';
export const FetchStudyCountFailureType =
  'datagateway_common:fetch_study_count_failure';
export const FetchStudyCountSuccessType =
  'datagateway_common:fetch_study_count_success';

export const FetchDownloadCartRequestType =
  'datagateway_common:fetch_download_cart_request';
export const FetchDownloadCartFailureType =
  'datagateway_common:fetch_download_cart_failure';
export const FetchDownloadCartSuccessType =
  'datagateway_common:fetch_download_cart_success';

export const AddToCartRequestType = 'datagateway_common:add_to_cart_request';
export const AddToCartFailureType = 'datagateway_common:add_to_cart_failure';
export const AddToCartSuccessType = 'datagateway_common:add_to_cart_success';

export const RemoveFromCartRequestType =
  'datagateway_common:remove_from_cart_request';
export const RemoveFromCartFailureType =
  'datagateway_common:remove_from_cart_failure';
export const RemoveFromCartSuccessType =
  'datagateway_common:remove_from_cart_success';

export const FetchAllIdsRequestType =
  'datagateway_common:fetch_all_ids_request';
export const FetchAllIdsFailureType =
  'datagateway_common:fetch_all_ids_failure';
export const FetchAllIdsSuccessType =
  'datagateway_common:fetch_all_ids_success';

export const FetchFilterRequestType = 'datagateway_common:fetch_filter_request';
export const FetchFilterFailureType = 'datagateway_common:fetch_filter_failure';
export const FetchFilterSuccessType = 'datagateway_common:fetch_filter_success';

export const IsisDatafileDetailsPanelChangeTabType =
  'datagateway_common:isis_datafile_details_panel_change_tab';
export const IsisDatasetDetailsPanelChangeTabType =
  'datagateway_common:isis_dataset_details_panel_change_tab';
export const IsisInstrumentDetailsPanelChangeTabType =
  'datagateway_common:isis_instrument_details_panel_change_tab';
export const IsisInvestigationDetailsPanelChangeTabType =
  'datagateway_common:isis_investigation_details_panel_change_tab';
export const DlsDatasetDetailsPanelChangeTabType =
  'datagateway_common:dls_dataset_details_panel_change_tab';
export const DlsVisitDetailsPanelChangeTabType =
  'datagateway_common:dls_visit_details_panel_change_tab';

export interface SortTablePayload {
  column: string;
  order: Order | null;
}

export interface FilterTablePayload {
  column: string;
  filter: Filter | null;
}

export interface ConfigureFacilityNamePayload {
  facilityName: string;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface FeatureSwitches {}

export interface ConfigureUrlsPayload {
  urls: URLs;
}

export interface URLs {
  idsUrl: string;
  apiUrl: string;
  downloadApiUrl: string;
  icatUrl: string;
}

export interface PluginRoute {
  section: string;
  link: string;
  displayName: string;
  admin?: boolean;
  hideFromMenu?: boolean;
  order: number;
}

export interface UpdateViewPayload {
  view: ViewsType;
}

export interface UpdateSearchPayload {
  search: string | null;
}

export interface UpdatePagePayload {
  page: number | null;
}

export interface UpdateResultsPayload {
  results: number | null;
}

export interface UpdateFiltersPayload {
  filters: FiltersType;
}

export interface UpdateSortPayload {
  sort: SortType;
}

export interface UpdateQueryPayload {
  query: QueryParams;
}

export interface SaveViewPayload {
  view: ViewsType;
}

export interface RequestPayload {
  timestamp: number;
}

export interface FailurePayload {
  error: string;
}

export interface FetchDataSuccessPayload {
  data: Entity[];
  timestamp: number;
}

export interface FetchDetailsSuccessPayload {
  data: Entity[];
}

export interface FetchIdsSuccessPayload {
  data: number[];
  timestamp: number;
}

export interface FetchFilterSuccessPayload {
  filterKey: string;
  data: string[];
}

export interface FetchCountSuccessPayload {
  count: number;
  timestamp: number;
}

export interface FetchDataCountSuccessPayload {
  id: number;
  count: number;
  timestamp: number;
}

export interface FetchSizeSuccessPayload {
  id: number;
  size: number;
}

export interface DownloadCartPayload {
  downloadCart: DownloadCart;
}

/**
 * Payload for the ISIS datafile details panel change tab action.
 */
export interface IsisDatafileDetailsPanelChangeTabPayload {
  datafileId: Datafile['id'];
  newTab: IsisDatafileDetailsPanelTab;
}

/**
 * Payload for the ISIS dataset details panel change tab action.
 */
export interface IsisDatasetDetailsPanelChangeTabPayload {
  datasetId: Dataset['id'];
  newTab: IsisDatasetDetailsPanelTab;
}

/**
 * Payload for the ISIS instrument details panel change tab action.
 */
export interface IsisInstrumentDetailsPanelChangeTabPayload {
  instrumentId: Instrument['id'];
  newTab: IsisInstrumentDetailsPanelTab;
}

/**
 * Payload for the ISIS investigation details panel change tab action.
 */
export interface IsisInvestigationDetailsPanelChangeTabPayload {
  investigationId: Investigation['id'];
  newTab: IsisInvestigationDetailsPanelTab;
}

/**
 * Payload for the DLS dataset details panel change tab action.
 */
export interface DlsDatasetDetailsPanelChangeTabPayload {
  datasetId: Dataset['id'];
  newTab: DlsDatasetDetailsPanelTab;
}

export interface DlsVisitDetailsPanelChangeTabPayload {
  investigationId: Investigation['id'];
  newTab: DlsVisitDetailsPanelTab;
}
