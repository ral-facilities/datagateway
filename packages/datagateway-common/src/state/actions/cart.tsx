import axios from 'axios';
import { Action } from 'redux';
import { getApiFilter } from '.';
import { DownloadCart, Investigation } from '../../app.types';
import handleICATError from '../../handleICATError';
import { readSciGatewayToken } from '../../parseTokens';
import { ActionType, ThunkResult } from '../app.types';
import {
  AddToCartFailureType,
  AddToCartRequestType,
  AddToCartSuccessType,
  DownloadCartPayload,
  FailurePayload,
  FetchAllIdsFailureType,
  FetchAllIdsRequestType,
  FetchAllIdsSuccessType,
  FetchDownloadCartFailureType,
  FetchDownloadCartRequestType,
  FetchDownloadCartSuccessType,
  FetchIdsSuccessPayload,
  RemoveFromCartFailureType,
  RemoveFromCartRequestType,
  RemoveFromCartSuccessType,
  RequestPayload,
} from './actions.types';

export const fetchDownloadCartSuccess = (
  downloadCart: DownloadCart
): ActionType<DownloadCartPayload> => ({
  type: FetchDownloadCartSuccessType,
  payload: {
    downloadCart,
  },
});

export const fetchDownloadCartFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: FetchDownloadCartFailureType,
  payload: {
    error,
  },
});

export const fetchDownloadCartRequest = (): Action => ({
  type: FetchDownloadCartRequestType,
});

export const fetchDownloadCart = (): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(fetchDownloadCartRequest());

    const { facilityName } = getState().dgcommon;
    const { downloadApiUrl } = getState().dgcommon.urls;

    await axios
      .get(`${downloadApiUrl}/user/cart/${facilityName}`, {
        params: {
          sessionId: readSciGatewayToken().sessionId,
        },
      })
      .then((response) => {
        dispatch(fetchDownloadCartSuccess(response.data));
      })
      .catch((error) => {
        handleICATError(error);
        dispatch(fetchDownloadCartFailure(error.message));
      });
  };
};

export const addToCartSuccess = (
  downloadCart: DownloadCart
): ActionType<DownloadCartPayload> => ({
  type: AddToCartSuccessType,
  payload: {
    downloadCart,
  },
});

export const addToCartFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: AddToCartFailureType,
  payload: {
    error,
  },
});

export const addToCartRequest = (): Action => ({
  type: AddToCartRequestType,
});

export const addToCart = (
  entityType: 'investigation' | 'dataset' | 'datafile',
  entityIds: number[]
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(addToCartRequest());

    const { facilityName } = getState().dgcommon;
    const { downloadApiUrl } = getState().dgcommon.urls;

    const params = new URLSearchParams();
    params.append('sessionId', readSciGatewayToken().sessionId || '');
    params.append(
      'items',
      `${entityType} ${entityIds.join(`, ${entityType} `)}`
    );

    await axios
      .post(`${downloadApiUrl}/user/cart/${facilityName}/cartItems`, params)
      .then((response) => {
        dispatch(addToCartSuccess(response.data));
      })
      .catch((error) => {
        handleICATError(error);
        dispatch(addToCartFailure(error.message));
      });
  };
};

export const removeFromCartSuccess = (
  downloadCart: DownloadCart
): ActionType<DownloadCartPayload> => ({
  type: RemoveFromCartSuccessType,
  payload: {
    downloadCart,
  },
});

export const removeFromCartFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: RemoveFromCartFailureType,
  payload: {
    error,
  },
});

export const removeFromCartRequest = (): Action => ({
  type: RemoveFromCartRequestType,
});

export const removeFromCart = (
  entityType: 'investigation' | 'dataset' | 'datafile',
  entityIds: number[]
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    dispatch(removeFromCartRequest());

    const { facilityName } = getState().dgcommon;
    const { downloadApiUrl } = getState().dgcommon.urls;

    await axios
      .delete(`${downloadApiUrl}/user/cart/${facilityName}/cartItems`, {
        params: {
          sessionId: readSciGatewayToken().sessionId,
          items: `${entityType} ${entityIds.join(`, ${entityType} `)}`,
        },
      })
      .then((response) => {
        dispatch(removeFromCartSuccess(response.data));
      })
      .catch((error) => {
        handleICATError(error);
        dispatch(removeFromCartFailure(error.message));
      });
  };
};

export const fetchAllIdsSuccess = (
  allIds: number[],
  timestamp: number
): ActionType<FetchIdsSuccessPayload> => ({
  type: FetchAllIdsSuccessType,
  payload: {
    data: allIds,
    timestamp,
  },
});

export const fetchAllIdsFailure = (
  error: string
): ActionType<FailurePayload> => ({
  type: FetchAllIdsFailureType,
  payload: {
    error,
  },
});

export const fetchAllIdsRequest = (
  timestamp: number
): ActionType<RequestPayload> => ({
  type: FetchAllIdsRequestType,
  payload: {
    timestamp,
  },
});

export const fetchAllIds = (
  entityType: 'investigation' | 'dataset' | 'datafile',
  additionalFilters?: {
    filterType: 'where' | 'distinct' | 'include';
    filterValue: string;
  }[]
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    const timestamp = Date.now();
    dispatch(fetchAllIdsRequest(timestamp));

    const params = getApiFilter(getState);
    if (additionalFilters) {
      additionalFilters.forEach((filter) => {
        params.append(filter.filterType, filter.filterValue);
      });
    }

    const distinctFilterString = params.get('distinct');
    if (distinctFilterString) {
      const distinctFilter: string | string[] = JSON.parse(
        distinctFilterString
      );
      if (typeof distinctFilter === 'string') {
        params.set('distinct', JSON.stringify([distinctFilter, 'id']));
      } else {
        params.set('distinct', JSON.stringify([...distinctFilter, 'id']));
      }
    } else {
      params.set('distinct', JSON.stringify('id'));
    }

    const { apiUrl } = getState().dgcommon.urls;

    await axios
      .get<{ id: number }[]>(`${apiUrl}/${entityType}s`, {
        params,
        headers: {
          Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
        },
      })
      .then((response) => {
        dispatch(
          fetchAllIdsSuccess(
            response.data.map((x) => x.id),
            timestamp
          )
        );
      })
      .catch((error) => {
        handleICATError(error);
        dispatch(fetchAllIdsFailure(error.message));
      });
  };
};

export const fetchAllISISInvestigationIds = (
  instrumentId: number,
  facilityCycleId: number
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    const timestamp = Date.now();
    dispatch(fetchAllIdsRequest(timestamp));

    const params = getApiFilter(getState);

    // TODO: currently datagateway-api can't apply distinct filter to ISIS queries,
    // so for now just retrieve everything
    // params.set('distinct', JSON.stringify('id'));

    const { apiUrl } = getState().dgcommon.urls;

    await axios
      .get<Investigation[]>(
        `${apiUrl}/instruments/${instrumentId}/facilitycycles/${facilityCycleId}/investigations`,
        {
          params,
          headers: {
            Authorization: `Bearer ${readSciGatewayToken().sessionId}`,
          },
        }
      )
      .then((response) => {
        dispatch(
          fetchAllIdsSuccess(
            response.data.map((x) => x.id),
            timestamp
          )
        );
      })
      .catch((error) => {
        handleICATError(error);
        dispatch(fetchAllIdsFailure(error.message));
      });
  };
};
