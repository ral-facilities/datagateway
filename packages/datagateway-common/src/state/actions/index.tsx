import { ActionType } from '../app.types';
import {
  ConfigureFacilityNamePayload,
  ConfigureFacilityNameType,
  ConfigureUrlsPayload,
  ConfigureURLsType,
  URLs,
} from './actions.types';

export const loadFacilityName = (
  name: string
): ActionType<ConfigureFacilityNamePayload> => ({
  type: ConfigureFacilityNameType,
  payload: {
    facilityName: name,
  },
});

export const loadUrls = (urls: URLs): ActionType<ConfigureUrlsPayload> => ({
  type: ConfigureURLsType,
  payload: {
    urls,
  },
});
