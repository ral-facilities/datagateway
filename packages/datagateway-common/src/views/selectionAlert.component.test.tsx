import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import { DownloadCartItem } from '../app.types';
import { NotificationType } from '../state/actions/actions.types';
import SelectionAlert from './selectionAlert.component';
import { render, RenderResult } from '@testing-library/react';
import { AnyAction } from 'redux';

describe('SelectionAlert', () => {
  let events: CustomEvent<AnyAction>[] = [];
  const cartItems: DownloadCartItem[] = [
    {
      entityId: 1,
      entityType: 'investigation',
      id: 1,
      name: 'Test',
      parentEntities: [],
    },
    {
      entityId: 2,
      entityType: 'dataset',
      id: 1,
      name: 'Test 2',
      parentEntities: [],
    },
    {
      entityId: 3,
      entityType: 'datafile',
      id: 2,
      name: 'Test 3',
      parentEntities: [],
    },
  ];
  let storageGetItemMock: jest.Mock;
  let storageSetItemMock: jest.Mock;
  let storageRemoveItemMock: jest.Mock;

  const createWrapper = (
    selectedItems: DownloadCartItem[],
    loggedInAnonymously: boolean
  ): ReactWrapper => {
    return mount(
      <SelectionAlert
        selectedItems={selectedItems}
        navigateToSelection={() => undefined}
        width={'100px'}
        marginSide={'4px'}
        loggedInAnonymously={loggedInAnonymously}
      />
    );
  };

  const createRTLWrapper = (
    selectedItems: DownloadCartItem[],
    loggedInAnonymously: boolean
  ): RenderResult => {
    return render(
      <SelectionAlert
        selectedItems={selectedItems}
        navigateToSelection={() => undefined}
        width={'100px'}
        marginSide={'4px'}
        loggedInAnonymously={loggedInAnonymously}
      />
    );
  };

  beforeEach(() => {
    events = [];
    document.dispatchEvent = (e: Event) => {
      events.push(e as CustomEvent<AnyAction>);
      return true;
    };

    storageGetItemMock = jest.fn();
    storageSetItemMock = jest.fn();
    storageRemoveItemMock = jest.fn();

    window.localStorage.__proto__.getItem = storageGetItemMock;
    window.localStorage.__proto__.setItem = storageSetItemMock;
    window.localStorage.__proto__.removeItem = storageRemoveItemMock;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const wrapper = createRTLWrapper([cartItems[0]], true);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('sends a notification to SciGateway if user is not logged in but only once', () => {
    const wrapper = createWrapper([cartItems[0]], true);

    wrapper.update();

    expect(events.length).toBe(1);
    expect(events[0].detail).toEqual({
      type: NotificationType,
      payload: {
        severity: 'warning',
        message: 'selec_alert.warning_message_session_token',
      },
    });

    expect(storageSetItemMock).toHaveBeenCalledTimes(1);
    expect(storageSetItemMock).toHaveBeenCalledWith('sentExpiredMessage', '1');

    storageGetItemMock.mockReturnValueOnce('1');

    //Adding another element
    wrapper.setProps({ selectedItems: [cartItems[0], cartItems[1]] });
    expect(events.length).toBe(1);

    storageGetItemMock.mockReturnValueOnce('1');

    //Removing all should reset the message
    wrapper.setProps({ selectedItems: [] });
    expect(events.length).toBe(1);

    expect(storageRemoveItemMock).toHaveBeenCalledTimes(1);
    expect(storageRemoveItemMock).toHaveBeenCalledWith('sentExpiredMessage');

    wrapper.setProps({ selectedItems: [cartItems[0]] });
    expect(events.length).toBe(2);
    expect(events[1].detail).toEqual({
      type: NotificationType,
      payload: {
        severity: 'warning',
        message: 'selec_alert.warning_message_session_token',
      },
    });
  });

  it('does not send a notification to SciGateway if user is not logged and it has been sent before', () => {
    storageGetItemMock.mockReturnValue('1');

    const wrapper = createWrapper([], true);

    wrapper.update();

    //Adding the first element should not broadcast a message
    wrapper.setProps({ selectedItems: [cartItems[0]] });
    expect(events.length).toBe(0);
  });

  it('renders correctly with more than one item selected', () => {
    const wrapper = createWrapper(cartItems, false);
    expect(
      wrapper.find('[aria-label="selection-alert-text"]').first().text().trim()
    ).toEqual('selec_alert.added');
  });

  it('renders correctly with one item removed', () => {
    const wrapper = createWrapper(cartItems, false);
    wrapper.setProps({ selectedItems: [cartItems[0], cartItems[1]] });

    expect(
      wrapper.find('[aria-label="selection-alert-text"]').first().text().trim()
    ).toEqual('selec_alert.removed');
  });

  it('renders correctly with all items removed', () => {
    const wrapper = createWrapper(cartItems, false);
    wrapper.setProps({ selectedItems: [] });

    expect(
      wrapper.find('[aria-label="selection-alert-text"]').first().text().trim()
    ).toEqual('selec_alert.removed');
  });

  it('does not render when nothing has changed', () => {
    const wrapper = createWrapper([], false);
    expect(wrapper.find('[aria-label="selection-alert"]').exists()).toBeFalsy();
  });

  it('does not render when closed', () => {
    const wrapper = createWrapper(cartItems, false);
    wrapper
      .find('[aria-label="selection-alert-close"]')
      .last()
      .simulate('click');
    wrapper.update();

    expect(wrapper.find('[aria-label="selection-alert"]').exists()).toBeFalsy();
  });

  it('clicking link calls navigateToSelection', () => {
    let navigated = false;
    const navigate = (): void => {
      navigated = true;
    };
    const wrapper = mount(
      <SelectionAlert
        selectedItems={cartItems}
        navigateToSelection={navigate}
        loggedInAnonymously={false}
      />
    );
    wrapper
      .find('[aria-label="selection-alert-link"]')
      .first()
      .simulate('click');

    expect(navigated).toBeTruthy();
  });
});
