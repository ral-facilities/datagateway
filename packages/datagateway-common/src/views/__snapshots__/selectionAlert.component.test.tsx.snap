// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionAlert renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="selection-alert"
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1wxid5k-MuiPaper-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-prmyfw-MuiGrid-root"
      >
        <p
          aria-label="selection-alert-text"
          class="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
        >
          selec_alert.added
        </p>
         
        <button
          aria-label="selection-alert-link"
          style="font-family: arial, sans-serif; color: rgb(4, 0, 145); text-decoration: underline; cursor: pointer; font-size: 14px;"
        >
          selec_alert.link
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          aria-label="selection-alert-close"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-1e0d89p-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
